### ISSUE TITLE
Need update method

### ISSUE THREAD

**Issue by inator** [hoodiehq/hoodie-task-server#38]
It just occurred to me that we may need an API method for updates, which would allow for the client to pick up progress (via task data) and such for a task that is being processed but not yet closed.

**Comment by mbad0la** [comment]
Hey @inator

Do let me know if I could be of any help!

**Comment by inator** [comment]
@mbad0la - I absolutely welcome the help if you're interested in taking this one on?  If so, I can share the implementation ideas @gr2m and I discussed today.  It's a fairly simple addition and I can get you oriented to the current API to get you started.  I'm currently focused on the task-client refactor, so your kind offer to help is greatly appreciated!

**Comment by inator** [comment]
We're thinking that a .progress() method would be more appropriate, with something like:

    var taskUpdate = { percentComplete: '45' }
    task.progress(taskDoc, taskUpdate)

Behind the scenes the taskDoc would have the taskUpdate value pushed into an array property named "progress", which in turn would be synced to the client for the developer to parse and use however they see fit, such as a progress ticker from the above example.

@gr2m, please chime in if this doesn't accurately represent our brainstorming.  Thanks!

**Comment by mbad0la** [comment]
I'd love to take this up. Just get me up to speed!

**Comment by inator** [comment]
@mbad0la - I DM'd you on Slack!  Thanks!

**Comment by inator** [comment]
@mbad0la - Gregor and I talked bit more on the progress method and have a slight change in approach:

    var options = { name: 'optional-name', progressData: data } 
    .progress(taskDoc, options)

Where the taskDoc is required and options.name and options.progressData are optional.  In the absence of options.name, it should default to 'progress'.  This should be pushed to the taskDoc.progress array property as follows:

    .progress: [{name: 'progress', at: '2017-01-03T21:46:55.688Z', progressData: data }]

options.progressData will contain what was perviously the 'progressValue'.

Sorry for the change midstream.  The tasks concept is maturing rather rapidly and subject to some adjustments. :)

**Comment by mbad0la** [comment]
> progress: [{name: 'progress', at: '2017-01-03T21:46:55.688Z', progressData: data }]

Will the server be generating the time for the `at` field (just before pushing the progress object into the `progress` array)?
Also, only the `options.name` is optional right? The `options.progressData` field would now be the necessary one isn't, instead of just the `progressData` argument discussed earlier?

**Comment by inator** [comment]
@mbad0la - yes you'll want to make that field server generated.

`options` and the both `options.name` and `options.progressData` are optional.  Absent `progressData` basically will supply a sort of "heart beat" back to the client letting it know the task is still processing.  In practice though, I can't imagine that devs will not include some `progressData` cargo to do something useful on the client with it - like tick up a progress bar.  Still, we wont error if they choose not to include it.

Make sense?

**Comment by inator** [comment]
Oops... need to update the api/README before we can close this. @mbad0la, you got that handled?

**Comment by mbad0la** [comment]
Sure @inator

Is there a documentation reference I can take an example from?

**Comment by inator** [comment]
@mbad0la - Minimally, if you just want to add the method reference to the existing [api/README](https://github.com/hoodiehq/hoodie-task-server/blob/master/api/README.md), that will fine for now.  I expect the docs for tasks will need some TLC before we go prime time with it all, so much of this is likely a placeholder.

Of course if you elect to go further with it, no harm in that either. :)  See the README for [hoodie-store-server-api](https://github.com/hoodiehq/hoodie-store-server-api/blob/master/README.md) for reference.

Thanks much!

**Comment by mbad0la** [comment]
I'll just add a reference for now. Would be better to have a separate PR for documentation.

**Comment by mbad0la** [comment]
Going through the docs, I found an anomaly in the code. According to the [docs](https://github.com/hoodiehq/hoodie-task-server/blame/master/api/README.md#L26) the API methods return a promise but I don't see that happening inside the `success` and `error` method. Please confirm.

**Comment by inator** [comment]
> Going through the docs, I found an anomaly in the code. According to the docs the API methods return a promise but I don't see that happening inside the success and error method. Please confirm.

Good catch!  I think they should... probably my error.  Just needs `return` added to each most likely.

**Comment by inator** [comment]
@mbad0la - thanks for handling this!  Closing.
